src/abstractarray.jl

```diff
@@ -126,9 +126,10 @@ similar(::Type{A}, shape::Tuple{SOneTo, Vararg{SOneTo}}) where {A<:AbstractArray
 similar(::A,::Type{T}, shape::Tuple{SOneTo, Vararg{SOneTo}}) where {A<:AbstractArray,T} = similar(A, T, Size(last.(shape)))
 similar(::Type{A},::Type{T}, shape::Tuple{SOneTo, Vararg{SOneTo}}) where {A<:AbstractArray,T} = similar(A, T, Size(last.(shape)))
 
-const SOneToLike{n} = Union{SOneTo{n}, Base.Slice{SOneTo{n}}}
+const SOneToLike{n} = Union{Integer, SOneTo{n}, IdentityUnitRange{SOneTo{n}}}
```

Comment:
I don't think I'd include `Integer` here — if you want to get a static array from an indexing slice, then I think you should have to use a static array of indices to do so.

Reply:
If I don't include it, the method `similar(::Type{A},::Type{T}, shape::Tuple{SOneToLike, Vararg{SOneToLike}})` in line 134 will not get dispatched to by the `similar` call from base (which is the origin of the error here), and the whole thing stops working. This seems to be the approach suggested in Base, I think. Check https://github.com/JuliaLang/julia/blob/e83693749032d28e9f5f0f75695fad46729ea2b4/base/abstractarray.jl#L581 and surrounding code/comments.

Reply:
Ah, right, of course.  We need dispatch on a Tuple with at least one `SOneToLike` to accomplish what we want here, which of course we cannot currently express (directly).

Here's what I think I'd do, in pseudocode:

```julia
const PossiblyStaticAxis = Union{Integer, OneTo, SOneTo, IdentityUnitRange{<:SOneTo}}
similar(…, ::Tuple{PossiblyStaticAxis, Vararg{PossiblyStaticAxis}}) = static_similar(…)

static_similar(…, ::Tuple{SOneTo, Vararg{SOneTo}}) = actually_static_similar(…)
static_similar(…, ::Tuple{Any, Vararg{Any}}) = similar(…, map(length, …))
```

I think that should do the trick — and hopefully make things work much better all around.  The method tables of `similar` are disastrously messy, so this is indeed quite hard.

Reply:
Yeah, I confess I struggle a bit getting my head around this :-). Thanks for your suggestion Matt. Feel free to take over this PR if you want. Otherwise I'll implement your suggestion tomorrow and report back.

Reply:
The problem with the modification you propose, as far as I can see, is that it only dispatches to `actually_static_similar` those shapes that are *all* `SOneTo`. The original PR first converts each `PossiblyStaticAxis` element of `shape` into an integer (the size along each direction), regardless of whether it is an `SOneTo`, and `IdentityUnitRange` or an actual `Int`. It then transforms it into a `Size` and proceeds. 
Another little problem is that I think `OneTo` needs not be a subtype of `PossiblyStaticAxis`, as it gets converted to and `Int` in base, before getting here. After playing with it for a while I've sort of convinced myself that the current PR is probably ok, without considerably reworking the design. 

One of the problems of the current design, it seems to me, is the difficulty of making it yield and `MArray` when `shape` is an `NTuple{Int}`, which is why `s[1:2, 1:2] isa Array` instead of returning an `MArray` as it should. If I try to add a specific dispatch for this, like 
```
similar(::SA, ::Type{T}, shape::Tuple{Integer, Vararg{Integer}}) where {SA<:StaticArray,T} = similar(SA, T, Size(shape))
```
I run into an ambiguity with the default `similar` method fallback in base `similar(a::AbstractArray, ::Type{T}, dims::Dims{N}) where {T,N} = Array{T,N}(undef, dims)`.